core/src/main/java/org/elasticsearch/common/xcontent/support/AbstractXContentParser.java

```diff
@@ -289,7 +289,7 @@ public BytesRef utf8BytesOrNull() throws IOException {
             token = parser.nextToken();
         }
         ArrayList<Object> list = new ArrayList<>();
-        for (; token != XContentParser.Token.END_ARRAY; token = parser.nextToken()) {
+        for (; token != null && token != XContentParser.Token.END_ARRAY && token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
```

Comment:
It is a small safeguard in case some code tries to read a list but only a value is available. `list()` and `readList()` methods should only be called when parsing an array but who knows.